### PR TITLE
Move to using GAS from YASM and improve system call code

### PR DESCRIPTION
**Pull request by graphitemaster**
This eliminates the use of YASM assembler files in favor of using the native assembler that comes with your toolchain. This also improves system call handling code by checking a few missed userspace passed pointers and file descriptors as well as improving branch prediction in some system call functions.

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -160,10 +164,11 @@ test: system
 toolchain:
 	@cd toolchain; ./toolchain-build.sh
 
+ASM_OBJS = kernel/boot.o kernel/gdt.o kernel/idt.o kernel/irq.o kernel/isr.o kernel/task.o kernel/tss.o kernel/user.o
```
Comment (klange) [inline Makefile]: I know you're trying to force an ordering on these so `boot` is first, but I'd much rather see this automatically generated like the other object lists. You're also duplicating effort from where you add `$(filter-out kernel/symbols.o,$(patsubst %.S,%.o,$(wildcard kernel/*/*.S)))` to `KERNEL_OBJS`.

File: kernel/cpu/gdt.c [hunk 2]
```diff
@@ -43,7 +43,7 @@ struct gdt_ptr		gp;
  * (ASM) gdt_flush
  * Reloads the segment registers
  */
-extern void gdt_flush(void);
+extern void gdt_flush(unsigned int);
```
Comment (klange) [inline kernel/cpu/gdt.c]: Should this be `uintptr_t`? Same for `idt`...

File: kernel/sys/syscall.c [hunk 3]
```diff
@@ -600,57 +618,55 @@ static int sys_gettimeofday(struct timeval * tv, void * tz) {
...
-	if (validate_safe(arg) || validate_safe(mountpoint) || validate_safe(type)) {
-		return -EFAULT;
-	}
-
-	if (current_process->user != USER_ROOT_UID) {
+	if (current_process->user != USER_ROOT_UID)
 		return -EPERM;
-	}
 
-	/* I may or may not start using these eventually. */
```
Comment (klange) [inline kernel/sys/syscall.c]: Should probably keep this comment or at least reformat it to a proper TODO:
```c
/* TODO: Make use of flags and data from mount command. */
```

File: kernel/cpu/gdt.c (hunk 2 above)
Comment (graphitemaster) [inline kernel/cpu/gdt.c]: could be uintptr_t if it's a 32-bit unsigned int, yeah.

File: Makefile (hunk 1 above)
Comment (graphitemaster) [inline Makefile]: ASM_OBJS could actually be generated kernel/boot.o followed by a glob of .S:.o + filter-out of kernel/boot.o, would that work for you?

File: kernel/sys/syscall.c (hunk 3 above)
Comment (graphitemaster) [inline kernel/sys/syscall.c]: Done

File: kernel/cpu/gdt.c (hunk 2 above)
Comment (graphitemaster) [inline kernel/cpu/gdt.c]: Done

File: Makefile (hunk 1 above)
Comment (graphitemaster) [inline Makefile]: Done

**Pull request merged**